File: CPP/Lesson_248_detect_squares.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 248 -- Detect Squares
// Category   : Math and Geometry
// Difficulty : Medium
// Study Plan : Day 124
// =============================================================
//
// QUESTION:
//   Add point or count axis-aligned squares with given query point.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class DS{map<pair<int,int>,int> c;public:void add(int x,int y){c[{x,y}]++;}int count(int x,int y){int tot=0;for(auto& [k,v]:c){auto [a,b]=k;if(a==x||abs(a-x)!=abs(b-y))continue;auto it1=c.find({a,y}),it2=c.find({x,b});if(it1!=c.end()&&it2!=c.end())tot+=v*it1->second*it2->second;}return tot;}};
int main(){DS d;d.add(3,10);d.add(11,2);d.add(3,2);cout<<d.count(11,10)<<"\n";}

File: Python/Lesson_248_detect_squares.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 248 -- Detect Squares
# Category   : Math and Geometry
# Difficulty : Medium
# Study Plan : Day 124
# =============================================================
#
# QUESTION:
#   Add point or count axis-aligned squares with given query point.
# =============================================================
class DS:
    def __init__(s): s.c={}
    def add(s,p): s.c[tuple(p)]=s.c.get(tuple(p),0)+1
    def count(s,p):
        x,y=p; tot=0
        for (a,b),v in list(s.c.items()):
            if a==x or abs(a-x)!=abs(b-y): continue
            tot += v*s.c.get((a,y),0)*s.c.get((x,b),0)
        return tot
if __name__=="__main__":
    d=DS(); [d.add(p) for p in [[3,10],[11,2],[3,2]]]; print(d.count([11,10]))
